### DISCUSSION TITLE
MKS DLC32 max support questions

### DISCUSSION THREAD

**Discussion by luc-github** [grblHAL/ESP32#132] [Ideas]
Hello I have bought a MKS DL32 Max board with its serial TFT
It is a ESP32S3 chip with 8MB flash - no PSRAM detected so far and no IO extender

Using https://github.com/makerbase-mks/MKS-DLC32/blob/main/MKS--DLC32-MAX-main/firmware/hardware.rar
[MKS DLC32 MAX V1.0_002 PIN.pdf](https://github.com/user-attachments/files/18217864/MKS.DLC32.MAX.V1.0_002.PIN.pdf)
and some tests I think am was able to list the global pinout :

|GPIO | Function | Description |
|-----|----------|-------------|
|GPIO0 | Flash mode | |
|GPIO1 | Air Pump |  Header|
|GPIO2 | Laser |  Header|
|GPIO3 | ?? | ??|
|GPIO4 | DIR Z | |
|GPIO5 | STEP Z | |
|GPIO6 | DIR Y1-Y2 | |
|GPIO7 | STEP Y1-Y2 | |
|GPIO8 | EN Stepper X Y1 Y2 Z A | |
|GPIO9 | ?? | ??|
|GPIO10 | SD Detect | |
|GPIO11 | SD D1 | SDIO D1|
|GPIO12 | SD D0 | SDIO D0|
|GPIO13 | SD CLK | |
|GPIO14 | SD CMD | |
|GPIO15 | DIR X | |
|GPIO16 | STEP X | |
|GPIO17 | TX1 | UART1 TX|   
|GPIO18 | RX1 | UART1 RX|
|GPIO19 | DIR A | |
|GPIO20 | STEP A | |
|GPIO21 | SD D3 | SDIO D3|
|GPIO33 | ?? | ??|
|GPIO34 | ?? | ??|
|GPIO35 | Spindle | |
|GPIO36 | ?? | ??|
|GPIO37 | ?? | ??|
|GPIO38 | ?? | ??|
|GPIO39 | END STOP X | also probe also A endstop|
|GPIO40 | END STOP Y | also Flame|
|GPIO41 | END STOP Z | also DOOR|
|GPIO42 | Beeper |  header|
|GPIO43 | TX0 | UART0 TX|
|GPIO44 | RX0 | UART0 RX|
|GPIO45 | ?? | ??|
|GPIO46 | ?? | ??|
|GPIO47 | SD D2 | SDIO D2|
|GPIO48 | ?? | ??|

the serial tft is using RX1/TX1

I have started to see how to build a mks_dlc32_max_1_0_map.h

About SD card how to specify the SD Detect pin ?Or I just ignore it ?
Is SDIO 4 bits Reader supported (CMD, CLK, D0, D1, D2, D3) ?Or should I try to map SPI pins instead ? (Mosi/Miso/SCK/CS)

here what board looks like
![image](https://github.com/user-attachments/assets/10540368-07d7-4fa9-bb9b-ea8a8bf40395)

the screen is using rj11 connector, but there is another connector on the side if needed 
![image](https://github.com/user-attachments/assets/33f98c9b-4c54-4521-8db0-162b90e2680d)

**Comment by terjeio** [comment]
I have just added [SDIO and SD card detect support](https://github.com/grblHAL/ESP32/commit/dcdcc7b4297590aa7b78b1f2389f8f3d3463ed64) to the driver:

```
#if SDCARD_ENABLE

#define SDCARD_SDIO
#define SD_CLK_PIN          GPIO_NUM_13
#define SD_CMD_PIN          GPIO_NUM_14
#define SD_DETECT_PIN       GPIO_NUM_10
#define SD_D0_PIN           GPIO_NUM_12
#define SD_D1_PIN           GPIO_NUM_11
#define SD_D2_PIN           GPIO_NUM_47
#define SD_D3_PIN           GPIO_NUM_21

#endif // SDCARD_ENABLE
```

Only tested with a breakout board and rather long cables, appeared to be unstable for me. Interested to hear if it works better with the board. And also if SDIO is significantly faster than SPI.

**Reply by luc-github** [reply]
excellent - thank you - I will give a try this week and feedback 

> [!NOTE]
>### I am late on this sorry, busy with new year and other project that take more time than expected,  still on it,  will feedback once done

**Comment by luc-github** [comment]
Update:
* I have created a file for the board according the pins and settings : [mks_dlc32_max_1_0_map.h](https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/boards/mks_dlc32_max_1_0_map.h)

* I have switched to idf 4.4.4 because in 4.4.3 S3 is premium so I was afraid not stable

* I have enabled the board in [my_machine.h](https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/my_machine.h) and put basic settings

* To make it compile I had to :
1 - rename sdkconfig to .old and rename sdkconfig.s3 to sdkconfig
2 - reconfigure for S3 : `idf.py set-target ESP32S3 reconfigure`
3 - be sure compilation follow USB_SERIAL_CDC so I have updated :
    * [driver.c](https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/driver.c#L69-L71)
    * [usb_serial.c](https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/usb_serial.c#L34)
4 - I2C pins are currently not defined/ used so I disabled it https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/boards/mks_dlc32_max_1_0_map.h#L120
     Note: EEPROM is disabled because it seems linked to I2C 
5 - Updated sdkconfig to use Single large APP partition because current one if overflowed - that is temporary because the flash is 8MB so can define a better partitions file

 Test Result:
 ```
 ?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1022|FS:0,0|Pn:XYZS|Ov:100,100,100|A:M>
ok
$I
[VER:1.1f.20250104:]
[OPT:VNMDSL,100,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,NOPROBE,ES,REBOOT,SED,RTC,WIFI,SD]
[FIRMWARE:grblHAL]
[SIGNALS:HSE]
[FREE MEMORY:205K]
[DRIVER:ESP32-S3]
[DRIVER VERSION:241222]
[DRIVER OPTIONS:v4.4.8-1-ga9d0f22193]
[WIFI MAC:64:e8:33:6e:a3:7c]
[IP:]
[PLUGIN:WebUI v0.24]
[PLUGIN:SDCARD v1.18]
[PLUGIN:FS macro plugin v0.13]
ok
$F
[FILE:/ANOLEX.NC|SIZE:18]
ok
```
SD is working and fast for what I can see, some comments:
- About SD
* if I remove the SD Card I got : `error:62` (failed to open dir)  but due to SD Detect pin you should get  `error:60` (failed to mount)
* if I insert SD Card I got debug output:
```
[0;32mI (789119) gpio: GPIO[13]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
[0;32mI (789119) gpio: GPIO[14]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
[0;32mI (789119) gpio: GPIO[12]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
[0;32mI (789129) gpio: GPIO[11]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
[0;32mI (789139) gpio: GPIO[47]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
[0;32mI (789149) gpio: GPIO[21]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 [0m
[0;32mI (789269) gpio: GPIO[21]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
```
- About Settings
doing $73, $74, $75  to configure STA does not stay if board if reboot , this is due to EEPROM disabled  because of I2C ? but on ESP32 you can use NVS - or I missed a setting or a command ? Sorry it has been a while I did not digged in grblHAL code

- About compilation
 you may add the sdkconfig path and partition path in CMakeList.txt, but I saw you moved out everything from it since last time I flashed grblHAL, so you may prefer people to rename sdkconfig instead
 
 
- About SERIAL1_PORT 

for the moment it seems not communicating with the serial screen, I did not found how to set the baudrate on it - it may be the issue ?


  
The board is not on a full system for the moment - I am just starting so tests are limited , let me know if you want me to check something 

Once done I can do a PR if you want

**Reply by terjeio** [reply]
> STA does not stay if board if reboot

Could it be [this bug](https://github.com/grblHAL/core/issues/659)? Other settings like $0 survive a reboot?

> if I remove the SD Card I got : error:62 (failed to open dir) but due to SD Detect pin you should get error:60 (failed to mount)

You should not get any output at all when removing a card. I have to wire up a dev. card again to test it.

> About SERIAL1_PORT
for the moment it seems not communicating with the serial screen, I did not found how to set the baudrate on it - it may be the issue ?

It could be the UART port and/or pins that are wrong? The USB port (J8) is listed as TXD0/RXD0 which it is certainly not so the screen port may be incorrect too? AFAIKT there is no schematic available so your only option is to try to follow the PCB traces to be sure of the pins?
The baudrate is [hardcoded here](https://github.com/grblHAL/ESP32/blob/d65b450e9bfff133ce185ecff4d122f2382a618e/main/driver.c#L3779), and note that there will be no output on the port unless enabled with [#define MPG_ENABLE 2](https://github.com/grblHAL/ESP32/blob/d65b450e9bfff133ce185ecff4d122f2382a618e/main/my_machine.h#L74) in my_machine.h. The MPG_STREAM symbol may be incorrect, should be 0 in the [stream_open_instance()](https://github.com/grblHAL/ESP32/blob/d65b450e9bfff133ce185ecff4d122f2382a618e/main/driver.c#L3779) when USB comms  is enabled.

> Once done I can do a PR if you want

Yes please.

**Reply by luc-github** [reply]
>>if I remove the SD Card I got : error:62 (failed to open dir) but due to SD Detect pin you should get error:60 (failed to mount)

>You should not get any output at all when removing a card. I have to wire up a dev. card again to test it.

Sorry I my explaination was not clear, I remove the SD and try to access it I get error:62 instead of error 60

the debug output appear when inserting the SD after boot

### about port 1:

I will investigate for the second port , I suspect pins are wrong indeed 

Edit: actually no, I tested pins with a script and they are corrects, may be it is more a communication issue with screen which use RJ11 connector - so I need to find some spare parts to debug , while I get those, I see the board has also a 4 pins serial connector that duplicate the RJ11 port - I will check if this one can communicate with grblHAL to check if blocking point is on mks tft and/or port itself 


### about settings not saved 

`$0=5.0`  does not survive to a restart
I am just using serial arduino monitor

**Reply by terjeio** [reply]
I see now why settings are lost, by changing the sdconfig to single app and not using the [partitions file](https://github.com/grblHAL/ESP32/blob/master/partitions.csv) there are no partition (the grbl one) for NVS storage available. You can see that in the `$I` report, this line will be present when NVS is available: `[NVS STORAGE:*FLASH]` and it is not there.

**Reply by terjeio** [reply]
> Sorry I my explaination was not clear, I remove the SD and try to access it I get error:62 instead of error 60

In the next commit error 64 will be returned for attempts to access the card when none is mounted.

**Reply by luc-github** [reply]
> I see now why settings are lost, by changing the sdconfig to single app and not using the [partitions file](https://github.com/grblHAL/ESP32/blob/master/partitions.csv) there are no partition (the grbl one) for NVS storage available. You can see that in the `$I` report, this line will be present when NVS is available: `[NVS STORAGE:*FLASH]` and it is not there.

Oh good catch, thank you - I was planning to redo the partitions.csv file for the 8MB flash - I will be sure this partition is present

> > Sorry I my explaination was not clear, I remove the SD and try to access it I get error:62 instead of error 60
> 
> In the next commit error 64 will be returned for attempts to access the card when none is mounted.

Ok noted - thank you 

Monday I will go to buy some spare parts to test the RJ11 communication and the MKS TFT screen and will report my findings

**Reply by luc-github** [reply]
@terjeio 

I have fixed the S3 SDCard output  behavior by updating sdkconfig to make the log to none and disabled the debub mode
I have updated also sdkconfig to support 8M flash and use a new partitions files to use th 8MB flash,
with new partition file the settings are now no more volatiles, thank you for your insight

I have update my_machine.h to enable MPG as you suggested and based on documentation on wiki
```
#define MPG_ENABLE              2 // Enable MPG interface. Requires a serial stream and means to switch between normal and MPG mode.
                                                    // 1: Mode switching is by handshake pin.
                                                    // 2: Mode switching is by the CMD_MPG_ENABLE_TOGGLE (0x8B) command character.
#define MPG_STREAM              0 // Enable MPG stream mode. Requires MPG_ENABLE.
#define KEYPAD_ENABLE           2 // 1: uses a I2C keypad for input.
                                    // 2: uses a serial stream for input. If MPG_ENABLE is set > 0 the serial stream is shared with the MPG.
```
I have added RX1/TX1 pins for serial1 in board file 
```
#define SERIAL1_PORT 
#define UART1_RX_PIN            GPIO_NUM_18
#define UART1_TX_PIN            GPIO_NUM_17
```

I have tested the pins with small arduino script to be sure can read and display
```
#include <Arduino.h>


void setup(){
    Serial1.begin(115200, SERIAL_8N1, 18, 17 );
}

void loop(){
    if (Serial1.available()){
        Serial1.print((char)Serial1.read());
    }
    delay(10);
}
```

I have added the support for the MPG toogle command in WebUI 3.0 because it was missing

I do not know if the above configuration is correct but behavior is not what I expected

* WebuI work as expected sending `$I`, `$$`, `?` command and get response as expected 
* Main serial  seems only supporting `?`, because `$$`, $I` give no answer in serial output  and no echo/response in Webui
* Serial1 : sending `?` do not give any response in serial1 output but answer is echo in webui and Main serial output, sending `$$`, `$I` do nothing, no response and no echo of response on others outputs

using 0x8B seems not changing anything, even not changing the report unlike documentation :  https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces suggest

Do I do a misconfiguration ?

**Reply by terjeio** [reply]
> I have added the support for the MPG toogle command in WebUI 3.0 because it was missing

The toggle command `0x8B` cannot be used from the WebUI, it is only available for [UART streams registered for MPG use](https://github.com/grblHAL/core/blob/c7c7d5a4cfcc47edd9995c235b50995d7a14f05a/stream.c#L487-L490) - and possibly later from a Bluetooth stream. If anything it could (or should?) disable its UI when `|MPG:1` is present in a realtime report and reenable it on `|MPG:0`.

> Serial1 : sending ? do not give any response in serial1 output but answer is echo in webui and Main serial output, sending $$, $I do nothing, no response and no echo of response on others outputs

The MPG stream should get a response from `?` either from itself or when sent from the WebUI. As long at is not in full control only realtime single command characters are responded to, regular commands (G-code or $-commands) are ignored. I'll have to check why no response is forthcoming.
If the keypad plugin is added to the build [many additional single character commands](https://github.com/grblHAL/Plugin_keypad/blob/master/README.md) are available that can be used for jogging etc. when the controller is in idle state and the MPG is not in full control.

FYI only one input stream can be in full control of grblHAL at any time, it is not possible to have two or more senders sending gcode at the same time.

> using 0x8B seems not changing anything, even not changing the report unlike documentation

A failed request to toggle MPG [mode on should result in a response](https://github.com/grblHAL/core/blob/c7c7d5a4cfcc47edd9995c235b50995d7a14f05a/stream.c#L540-L544), again I'll have to check why not.

**Reply by luc-github** [reply]
>If the keypad plugin is added to the build [many additional single character commands](https://github.com/grblHAL/Plugin_keypad/blob/master/README.md) are available that can be used for jogging etc. when the controller is in idle state and the MPG is not in full control.

I have just followed https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces#uart-interface
so you mean I should not do `#define KEYPAD_ENABLE 2` ? sorry I am confused

**Reply by terjeio** [reply]
Oops, you need to enable [MPG mode](https://github.com/grblHAL/ESP32/blob/a8032788d1c5738d64cd7fa28605cfe0630e18ac/main/my_machine.h#L74-L78) as well  (`#define MPG_ENABLE 2`) to get any output on the port - the keypad plugin only handles input.

**Reply by luc-github** [reply]
yes this is what I did :
https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/my_machine.h#L75-L79

**Reply by terjeio** [reply]
Hmm, if you have USB_CDC enabled the MPG stream will be the primary UART since by default the first unclaimed stream is assigned to the MPG. This could be the reason you do not get any output.
To force the MPG stream to always use serial1 add `#define MPG_STREAM 1` to the map file. For now you will have to add `#define KEYPAD_STREAM 1` as well to have the MPG and keypad share the same stream. I'll change that to share the streams if the keypad stream is not explicitly defined to use a different stream.

When I did that MPG mode almost works... If I switch to MPG mode before connecting via Telnet or Websocket (WebUI) the resident guru awakens and reboots the controller, why I do not know. By delaying the transmission of the status report output on a switchover by 5ms I managed to not disturb him so I will add that change in the next commit. Switching MPG mode after the connection has been made does not require the delay so it might be related to RTOS task switching since I cannot remember seeing this with other MCUs.

FYI if MPG mode is active and a Telnet or Websocket connection is made MPG mode is switched off for the duration of the connection unless the MPG application rerequest MPG mode to regain control.

Tip: if you want you can change the `0x8B` command character to a printable character for test purposes, this allows one to use a terminal program to simulate a MPG. I typically [replace 0x8B](https://github.com/grblHAL/core/blob/f4ca082f38f8383fb1e7e78fad8b75fcbb118be6/grbl.h#L111) with `'&'`.

**Reply by luc-github** [reply]
>if you have USB_CDC enabled the MPG stream will be the primary UART since by default the first unclaimed stream is assigned to the MPG. This could be the reason you do not get any output.

no I did `#define USB_SERIAL_CDC          0 `, or you mean better no define at all ?

>FYI if MPG mode is active and a Telnet or Websocket connection is made MPG mode is switched off for the duration of the connection unless the MPG application rerequest MPG mode to regain control.

Ok so the screen should be able to send such request / command, noted

>Tip: if you want you can change the 0x8B command character to a printable character for test purposes, this allows one to use a terminal program to simulate a MPG. I typically [replace 0x8B](https://github.com/grblHAL/core/blob/f4ca082f38f8383fb1e7e78fad8b75fcbb118be6/grbl.h#L111) with '&'

Yes saw it but because I initially though that was a global command and not input specific I have added the command to WebUI, now I understand it is input specific as you explained: 
>The toggle command 0x8B cannot be used from the WebUI, it is only available for [UART streams registered for MPG use](https://github.com/grblHAL/core/blob/c7c7d5a4cfcc47edd9995c235b50995d7a14f05a/stream.c#L487-L490)

 so I will test changing the char 
 
 Thank you for your help I will test and report

**Reply by luc-github** [reply]
@terjeio 
I have updated source with latest code and followed your suggestions:
https://github.com/luc-github/ESP32/blob/mks_dlc32_max/main/my_machine.h#L75-L79

I have actually disabled keypad because I do not have one, I initially followed the documentation that wrote need to enable it.

Now it is working as expected :
* Serial is initially on control
* WebSocket take the control if webUI is open and release it when page is closed
* On Serial1 no control by default, only take control if use the toggle command - I have tested with `&` as you suggested
* If WebUI is connected, Serial and Serial 1 are read only, but using `&` on serial 1 still allow serial1 to send command
* Using `&` on Serial has no effect

About that last statement I have questions:  
- Why Serial cannot send the toogle command ? it is a toogle of a state, origin of command should not matter
   If the screen is not responding for any reason when it is in control - currently there is no way for Serial to recover the control but restart the board or open and use WebUI- is that an acceptable behavior ?  knowing that ? is always working what ever the origin and the state of MPG
- Is that normal the MPG report is only when MPG state change ?  not periodicaly ?
```
>&
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|WCO:0.000,0.000,0.000|WCS:G54|A:M|Sc:|MPG:1|H:0|T:0|TLR:0|FW:grblHAL>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1022|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1022|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1021|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1022|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|WCO:0.000,0.000,0.000>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|Ov:100,100,100|A:M>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1021|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
ok
>&
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|MPG:0>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|WCO:0.000,0.000,0.000>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|Ov:100,100,100|A:M>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|WCO:0.000,0.000,0.000>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS|Ov:100,100,100|A:M>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
>?
<Idle|MPos:0.000,0.000,0.000|Bf:100,1023|FS:0,0|Pn:XYZS>
```

**Reply by terjeio** [reply]
> If WebUI is connected, Serial and Serial 1 are read only, but using & on serial 1 still allow serial1 to send command

Serial and Serial 1 can send real time commands at any time and they will be responded to. In addition Serial 1 can send single character commands to the keypad plugin that will be expandend and acted upon if the controller is in a state allowing that.

> Why Serial cannot send the toogle command ? it is a toogle of a state, origin of command should not matter

Actually [it can](https://github.com/grblHAL/core/blob/dfa1b262a41d711bd3d2c09fb8f11ce43679afc1/protocol.c#L948-L951),  but only to turn MPG mode off. Allowing any client to turn it on can be implemented but then the sender should somehow be made aware that a MPG is available?

> Is that normal the MPG report is only when MPG state change ? not periodicaly ?

Yes, because the MPG should normally request status reports when in control. If not auto reporting can be enabled with $481 (auto report interval). There is a toggle [command character](https://github.com/grblHAL/core/blob/dfa1b262a41d711bd3d2c09fb8f11ce43679afc1/grbl.h#L112) for turning that on/off as well since it may mess with senders or even MPGs... Changes to the auto report mode will be added to the real-time report so it can be tracked.
FYI ioSender checks for auto reporting beeing enabled on connect and turns it off while connected since it has to be in control of when reports are needed.

So handling a MPG can be quite complicated - MPG implementations may be different and senders may or may not be able to handle switchover nicely etc. etc.

**Reply by luc-github** [reply]
thank for the clarifications, yes handling a MPG can be quite complicated, I agree

I will do the PR this week then and let you review it
I will check how to add the toogle command support to the MKS screen, but it will be next month , now here it is chinese new year, it is tonight actually, so all MKS teams are off, and I need to ask them because current FW source code is not yet available

**Comment by luc-github** [comment]
I saw there were some update on pinout 2 weeks ago (https://github.com/makerbase-mks/MKS-DLC32/blob/main/MKS--DLC32-MAX-main/firmware/MKS%20DLC32%20MAX%20V1.0_002%20PIN.pdf), I didn't tested these pins before but new pinout make sense for me.
 
Door: GPIO41->GPIO38
Flame: GPIO40 ->GPIO36
Probe: GPIO39->GPIO37
Axis A: GPIO39->GPIO48

I am now asking MKS if this change is because of a new board revision or a typo in original pdf, once I got the feedback, I will push a pr  to update pins accordingly in map file

I have also asked how to add grblHAL support for the serial companion screen =>TBD
